File: modules/expo-exo-player/src/ExpoExoPlayerModule.ts
import { NativeModule, requireNativeModule } from "expo";

import type { OnPlayerErrorEvent, PlayerState, onPlayerStatusUpdateEvent } from "@/typings/player";

type ExpoExoPlayerModuleEvents = {
  onPlayerStatusUpdate: (event: onPlayerStatusUpdateEvent) => void;
  onPlayerError: (event: OnPlayerErrorEvent) => void;
};

declare class ExpoExoPlayerModule extends NativeModule<ExpoExoPlayerModuleEvents> {
  prepare(file: string): void;
  release(): void;
  setPlay(): void;
  setPause(): void;
  seekTo(position: number): void;
  getPlayerState(): PlayerState;
}

export default requireNativeModule<ExpoExoPlayerModule>("ExpoExoPlayer");

File: modules/expo-exo-player/android/src/main/java/expo/modules/exoplayer/ExpoExoPlayerModule.kt
package expo.modules.exoplayer

import android.content.Context
import android.os.Bundle
import androidx.core.net.toUri
import androidx.core.os.bundleOf
import androidx.media3.common.MediaItem
import androidx.media3.common.PlaybackException
import androidx.media3.exoplayer.ExoPlayer
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoExoPlayerModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("ExpoExoPlayer")

        Function("prepare") { file: String ->
            val media = MediaItem.fromUri(file.toUri())
            player.setMediaItem(media)
            player.prepare()
        }

        Function("release") {
            player.release()
        }

        Function("setPlay") {
            player.play()
        }

        Function("setPause") {
            player.pause()
        }

        Function("isPlaying") {
            return@Function player.isPlaying
        }

        Function("seekTo") { position: Long ->
            player.seekTo(position)
        }

        Function("getCurrentPosition") {
            return@Function player.currentPosition
        }

        Function("getDuration") {
            return@Function player.duration
        }

        Events("onPlaybackStateChanged")

        Events("onPlayerError")

        OnStartObserving {
            player.addListener(listener)
        }

        OnStopObserving {
            player.removeListener(listener)
        }
    }

    private val ctx get(): Context = requireNotNull(appContext.reactContext)
    private val player get() = requireNotNull(ExoPlayer.Builder(ctx).build())
    private val listener = ExpoExoPlayerListener(this)

    fun onPlaybackStateChangedAsBundle(state: Int): Bundle {
        return bundleOf("state" to state)
    }

    fun onPlayerErrorAsBundle(error: PlaybackException): Bundle {
        return bundleOf("error" to error)
    }
}
